cmd/cloud-controller-manager/nodeipamcontroller.go

```diff
@@ -131,24 +119,12 @@ func startNodeIpamController(ccmConfig *cloudcontrollerconfig.CompletedConfig, n
 		}
 	}
 
-	var nodeCIDRMaskSizeIPv4, nodeCIDRMaskSizeIPv6 int
```

Comment:
Here you are merging three functions: `setNodeCIDRMaskSizesDualStack`, `setNodeCIDRMaskSizes` and `getNodeCIDRMaskSizes`.

I guess that it should be enough to just validate passed flags and merge `setNodeCIDRMaskSizesDualStack` with `setNodeCIDRMaskSizes` (Throw error if mixed flags are provided) if we only want to remove `features.IPv6DualStack`.

However, it seems that more is happening here. What we are actually doing is we are adding new validation that prevents running controllers with `--node-cidr-mask-size-ipv4` or `--node-cidr-mask-size-ipv6` flags if we are in single stack env (as, `clusterCIDRs` contains single IP). Previously this behavior was to ignore additional flags, now we are changing to a more strict behavior and throw errors. Is it by design?

Reply:
> Previously this behavior was to ignore additional flags

it wasn't ignore them, it always failed if you use the wrong flags for the current ip family mode, where do you see it?

> Is it by design?

I copied it directly from current kubernetes code, this syncs the controller validation

https://github.com/kubernetes/kubernetes/blob/85524e7da6f3895a97c1502f6d5d2e28c5c66de4/cmd/cloud-controller-manager/nodeipamcontroller.go#L126

Reply:
> I copied it directly from current kubernetes code, this syncs the controller validation
> 
> https://github.com/kubernetes/kubernetes/blob/85524e7da6f3895a97c1502f6d5d2e28c5c66de4/cmd/cloud-controller-manager/nodeipamcontroller.go#L126

I don't think anyone uses cmd/cloud-controller-manager/nodeipamcontroller.go from k/k.

Reply:
> Previously we would ignore additional flag: [setNodeCIDRMaskSizesDualStack](https://github.com/kubernetes/cloud-provider-gcp/blob/57686c3ef8a190d783710065d929245fbb888b3e/cmd/cloud-controller-manager/nodeipamcontroller.go#L230) would return some values(potentially default), but getNodeCIDRMaskSizes would only use single value that matches existing [ClusterCIDRs[0]](https://github.com/kubernetes/cloud-provider-gcp/blob/57686c3ef8a190d783710065d929245fbb888b3e/cmd/cloud-controller-manager/nodeipamcontroller.go#L235).

That seems an overlook during the feature development that got fixed now

> I don't think anyone uses cmd/cloud-controller-manager/nodeipamcontroller.go from k/k.

cloud-controller-manager from k/k is copied from kube-controller-manager AFAIK https://github.com/kubernetes/kubernetes/blob/85524e7da6f3895a97c1502f6d5d2e28c5c66de4/cmd/kube-controller-manager/app/core.go#L625 and I know for sure that is being used and is the base of this controller

Reply:
> That seems an overlook during the feature development that got fixed now

But now it's a change to a stricter value validation which is potentially a breaking change. I can imagine that one is setting IPv4 and IPv6 flags for all clusters in their scripts, and only ClusterCIDRs are used to decide which values will be used.

>  I know for sure that is being used and is the base of this controller
Yes, KCM's nodeipam is used. But CCM's nodeipam - not really. Is there any difference between KCM's nodeipam and CCM's nodeipam in K/K?

Reply:
> But now it's a change to a stricter value validation which is potentially a breaking change. I can imagine that one is setting IPv4 and IPv6 flags for all clusters in their scripts, and only ClusterCIDRs are used to decide which values will be used.
> 

ok, let me relax the validation then, I still feel is wrong, but for keeping compatibility you are right

> Yes, KCM's nodeipam is used. But CCM's nodeipam - not really. Is there any difference between KCM's nodeipam and CCM's nodeipam in K/K?

same controller, duplicate startup script

Reply:
I guess we can make this breaking change when bumping to 1.27. I don't want to carry any nodeipam patches, but making a breaking change in flags without any minor release seems just wrong.

Let's add this stricter validation when moving to 1.27.

SG?

Reply:
new patch to soft the validation again PRAL